Document Sec-CH-Prefers-Color-Scheme Client Hint header (#22266)
mdn/content: files/en-us/web/http/headers/sec-ch-save-data/index.md

---
title: Sec-CH-Save-Data
slug: Web/HTTP/Headers/Sec-CH-Save-Data
tags:
  - Sec-CH-Save-Data
  - Client hint
  - HTTP
  - HTTP Header
  - Reference
  - Request header
  - Experimental
browser-compat: http.headers.Sec-CH-Save-Data
---

{{HTTPSidebar}}{{SeeCompatTable}}{{SecureContext_Header}}

> The **`Sec-CH-Save-Data`** [user agent client hint](/en-US/docs/Web/HTTP/Client_hints#user_preference_media_features_client_hints) request header indicates the user agent's preference for reduced data usage. This could be for reasons such as high transfer costs, slow connection speeds, etc.

Reviewer: ```suggestion
The **`Sec-CH-Save-Data`** [user agent client hint](/en-US/docs/Web/HTTP/Client_hints#user_preference_media_features_client_hints) request header indicates the user agent prefers less data be sent. This could be for reasons such as high transfer costs, slow connection speeds, and so on.
```

> 

Reviewer: I think this is a low entropy hint - you will need to check, and if so add some text similar to what is in `Save-Data` description, and also link to it from https://developer.mozilla.org/en-US/docs/Web/HTTP/Client_hints#low_entropy_hints

What makes this a little complicated is that you don't have to request a low entropy hint - if it is supported it gets sent by default. That means 
- you might not have the request cycle as below except to indicate that you want the value to vary.
- You wouldn't need the cycle below because it exists to ensure that the header gets sent before rendering - if it is low entropy that will always nappen.

Further, low entropy hints consume bandwidth - you wouldn't want to send both Save-Data and Sec-CH-Save-Data. So we need to find out how the spec handle this - are they both being sent?

> If a server signals to a client via the {{httpheader("Accept-CH")}} header that it accepts `Sec-CH-Save-Data`, the client can then respond with this header to indicate the user's preference for reduced data usage. The server can send the client appropriately adapted content, for example, JavaScript, videos, images, or CSS, to reduce the amount of data transfered for displaying subsequent rendered content. This could include reducing the amount of images, videos, simpler CSS styles, simpler HTML markup. Further, `Sec-CH-Save-Data` should be used to reduce data sent to the client irrespective of the values of other client hints that indicate network capability, like {{HTTPHeader("Downlink")}} and {{HTTPHeader("RTT")}}.

Reviewer: 1. Text here repeats itself
2. If this is a low entropy hint you don't need it in Accept-CH

```suggestion
On recieving this header in a request, a server can send the client appropriately adapted content.
For example, it can send fewer and lower resolution images and videos, and simpler CSS styles or HTML markup. 

Note that  `Sec-CH-Save-Data` should be used to reduce data sent to the client irrespective of the values of other client hints that affect network capability, such as {{HTTPHeader("Downlink")}} and {{HTTPHeader("RTT")}}.
```

<table class="properties">
  <tbody>
    <tr>
      <th scope="row">Header type</th>
      <td>
        {{Glossary("Request header")}},
        <a href="/en-US/docs/Web/HTTP/Client_hints">Client hint</a>
      </td>
    </tr>
    <tr>
      <th scope="row">{{Glossary("Forbidden header name")}}</th>
      <td>yes</td>
    </tr>
  </tbody>
</table>

## Syntax

```http
Sec-CH-Save-Data: <boolean>
```

### Directives

- `<boolean>`

>   - : `?1` indicates that the user-agent prefers experience with reduced data usage (`true`).
>     `?0` indicates that user-agent prefers a full-featured experience, potentially with data-intensive components (`false`).

Reviewer: ```suggestion
  - : `?1` indicates that the user-agent prefers that less data be sent (`true`).
    `?0` indicates that the user-agent prefers the "normal data" version of the site (`false`).
```

## Examples

The client makes an initial request to the server:

```http
GET / HTTP/1.1
Host: example.com
```

> The server responds, telling the client via {{httpheader("Accept-CH")}} that it accepts `Sec-CH-Save-Data`. In this example {{httpheader("Critical-CH")}} is also used, indicating that `Sec-CH-Save-Data` is considered a [critical client hint](/en-US/docs/Web/HTTP/Client_hints#critical_client_hints).

Reviewer: FWIW Would you actually make this Critical-CH in real life? It is ironic to apply the header as critical because you've already had the response that was not "low data" - so despite the user saying they wanted to save data you're now asking for the file again ... albeit in a lower data form.

I guess the argument for doing so is that the user won't see a significant change in quality on the next page they go to in the same session. I'm not completely convinced.

Anyway, I guess I'm saying "as a general rule", not every hint should be a critical hint, and not every example needs to demonstrate this. Something to think about.

Author: Yes, I would consider making `Sec-CH-Save-Data` a critical hint and customizing page markup for low-data usage. For example [this blog](https://css-tricks.com/help-users-save-data/) writes about using `Save-Data` header and removing images, gifs and CSS styles from the markup entirely to save data. As far as I understand, back then there was no equivalent of `Critical-CH` so the browser would download the page twice (once requesting default full-featured experience and then immediately would reload requesting data-saving one). However, as far as I understand, `Critical-CH` enables server to avoid extra data transfers at the expense of extra round trips: if server sees a request without `Sec-CH-Save-Data`, it responds with `Accept-CH`, `Vary`, and `Critical-CH` and some "dummy" content (even just a spinner), then client sees `Critical-CH` and makes a new request with `Sec-CH-Save-Data` and server can now respond with customized markup.

It's possible to prepare three pre-baked responses (1) when header is not received (which uses media queries, including the upcoming `prefers-reduced-data`, image source sets, lazy loading, etc.), (2) data-saving mode (which removes almost everything, including unnecessary images) and (3) full experience which has lots of images defined in source sets, videos, media preloading, etc.

Reviewer: @bershanskiy I don't think it makes sense to do so - see my comment https://github.com/mdn/content/pull/22266/files#r1027381368

Reviewer: Depends what the spec says - it if is not low entropy, then that's fine.

```http
HTTP/1.1 200 OK
Content-Type: text/html
Accept-CH: Sec-CH-Save-Data

> Vary: Sec-CH-Prefers-Color-Scheme

Reviewer: ```suggestion
Vary: Sec-CH-Save-Data
```

Critical-CH: Sec-CH-Save-Data
```

> **Note:** We've also specified `Sec-CH-Save-Data` in the {{httpheader("Vary")}} header to indicate to the browser that the served content will differ based on this header value, even if the URL stays the same, so the browser shouldn't just use an existing cached response and instead should cache this response separately. Each header listed in the `Critical-CH` header should also be present in the `Accept-CH` and `Vary` headers.

> The client automatically retries the request (due to `Critical-CH` being specified above), telling the server via `Sec-CH-Save-Data` that it has a user preference for data-reduced-motion animations:

Reviewer: ```suggestion
The client automatically retries the request (due to `Critical-CH` being specified above), telling the server via `Sec-CH-Save-Data` that it has a user preference for reduced data to be sent.
```
